### PR TITLE
Implement Lightblue document event repository

### PR DESCRIPTION
**Pull request by alechenninger**
Previously there was LightblueEventRepository, now there is LightblueDocumentEventRepository and LightblueNotificationRepository. This is the implementation and tests for LightblueDocumentEventRepository only.

I can start working on error handling scenarios, but I figured I'll put that in another PR. If you want that in this PR also, I can do that. It's purely what works better for review, and also so we can start looking this over now and I can keep working on something else while you do.

### CODE REVIEW

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        this.clock = clock;
+
+        locking = lightblue.getLocking(lockingDomain);
+    }
+
+    @Override
+    public List<Notification> retrieveOldestNotificationsUpTo(int maxEvents)
+            throws LightblueException {
+        try {
+            // TODO: Either block until lock acquired or throw exception and let caller poll
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Methinks this is done, no?

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: Yep! Although it was somewhat of a question, actually. Should this method block, polling until the lock is acquired (with some timeout?) or should the caller deal with polling this method?

It's not a huge deal either way. If this method throws then it just needs to be explicitly documented that that is a reason it can throw. The tradeoff is more about implementations having to worry about polling until they can acquire lock or if they can fail fast and let caller deal with it.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            NotificationEntity[] notificationEntities = lightblue
+                    .data(FindRequests.newNotificationsForEntitiesUpTo(entities, maxEvents))
+                    .parseProcessed(NotificationEntity[].class);
+
+            lightblue.data(UpdateRequests.notificationsAsProcessing(notificationEntities));
+
+            return Arrays.stream(notificationEntities)
+                    .map(entity -> notificationFactory.getNotificationForEntity(entity, requester))
+                    .collect(Collectors.toList());
+        } finally {
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: I just opened lightblue-platform/lightblue-client#229 for try-with-resources support. Just an idea.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/UpdateRequests.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            List<Update> updates = new ArrayList<>(2);
+            updates.add(Update.set("status", entity.getStatus().toString()));
+
+            ZonedDateTime processedDate = entity.getProcessedDate();
+
+            if (processedDate != null) {
+                updates.add(Update.set("processedDate", Date.from(processedDate.toInstant())));
+            }
+
+            // Work around client bug.
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/UpdateRequests.java]: Can you please reference the issue in the comment?

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java (hunk 2 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: I actually experimented a bit with this originally but decided it was not necessary at the moment.

A locking abstraction will come in handy when we correctly deal with TTLs and "reacquiring" when it is still needed. That may be a good time to use try-with-resources.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/UpdateRequests.java (hunk 3 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/UpdateRequests.java]: Good call

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java (hunk 1 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueNotificationRepository.java]: This is updated in other revision (which works on the notification repository)

**Comment by kahowell** [comment]
Still haven't referenced the issue above...

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * Supports creating many types of {@link LightblueDocumentEvent}s based on the canonical message
+ * type in the provided {@link DocumentEventEntity}.
+ */
+public class ByTypeDocumentEventFactory implements DocumentEventFactory {
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java]: This is not much more than a map of `DocumentEventFactories`. So I'd like to propose we rename to `DocumentEventFactoryMap` or `DocumentEventFactoryTypeMap`. Thoughts?

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java]: I dig either of those. Out of curiosity, either of these appeal to you?

- `DocumentEventFactoriesByType`
- `DocumentEventFactoryMapByType`
- `DocumentEventFactoryByType`

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/FindRequests.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.redhat.lightblue.client.Literal;
+import com.redhat.lightblue.client.Projection;
+import com.redhat.lightblue.client.Query;
+import com.redhat.lightblue.client.Sort;
+import com.redhat.lightblue.client.request.data.DataFindRequest;
+
+import org.esbtools.eventhandler.lightblue.model.DocumentEventEntity;
+
+public abstract class FindRequests {
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/FindRequests.java]: At the risk of sounding like a cliche, I think we could move these classes into a util package... thoughts? Either way, I'd like to see them outside of the general package (related, I generally thing that package could be split up a bit). Thoughts?

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/FindRequests.java]: Generally I try to keep package hierarchies as flat as possible until one really seems useful. I'd be okay with a "requests" package perhaps? Util is too generic / "catch all" IMHO

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+            // We have a new event, let's see if it is superseded by or can be merged with any
+            // previous events we parsed or created as a result of a previous merge.
+            while (optimizedIterator.hasNext()) {
+                LightblueDocumentEvent previouslyOptimizedEvent = optimizedIterator.next();
+
+                if (newEvent.isSupersededBy(previouslyOptimizedEvent)) {
+                    // Keep previous event...
+                    DocumentEventEntity previousEntity = previouslyOptimizedEvent.wrappedDocumentEventEntity();
+                    previousEntity.addSurvivorOfIds(newEventEntity.getSurvivorOfIds());
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Why is this line necessary? If the event is new, it shouldn't have any `survivorOfIds` on it?

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                    newEventEntity.setStatus(DocumentEventEntity.Status.merged);
+                    newEventEntity.setProcessedDate(ZonedDateTime.now(clock));
+                    entitiesToUpdate.add(newEventEntity);
+
+                    // We create a new event as a result of the merger, and keep this instead of the
+                    // others.
+                    LightblueDocumentEvent merger = newEvent.merge(previouslyOptimizedEvent);
+                    DocumentEventEntity mergerEntity = merger.wrappedDocumentEventEntity();
+                    mergerEntity.addSurvivorOfIds(previousEntity.getSurvivorOfIds());
+                    mergerEntity.addSurvivorOfIds(newEventEntity.getSurvivorOfIds());
```
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Same ? as above

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java (hunk 4 above)
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java]: `DocumentEventFactoryMapByType` is pretty nice. The others feel like factory function names.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/FindRequests.java (hunk 5 above)
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/FindRequests.java]: I'd prefer requests package, but okay with leaving as-is too.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 6 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: The new event may be one that was already created as a result of a merge, therefore it will have already survived some events. We want to make sure we track those as well.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 7 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Same answer

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java (hunk 6 above)
Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Can you add a comment to that effect?

Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: How about renaming `newEvent` and `newEventEntity` to `newOrMergerEvent` and `newOrMergerEventEntity` ?

Comment (kahowell) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java]: Sounds good to me.

File: lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java (hunk 4 above)
Comment (alechenninger) [inline lightblue/src/main/java/org/esbtools/eventhandler/lightblue/ByTypeDocumentEventFactory.java]: What do you think about just inlining this class then?

That is, pass the document event repository a map of types to document event factories and remove this class?

**Comment by kahowell** [comment]
LGTM. Let's merge it in soon.
